enable "with X():" behavior for torch dispatch and function mode

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78215
* __->__ #78214

Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

Commit:
enable with semantics

[ghstack-poisoned]

Commit:
Update on "enable "with X():" semantics for torch dispatch and torch function mode"


Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

[ghstack-poisoned]

Commit:
Update on "enable "with X():" semantics for torch dispatch and torch function mode"


Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

[ghstack-poisoned]

Commit:
Update on "enable "with X():" semantics for torch dispatch and torch function mode"


Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

[ghstack-poisoned]

Commit:
Update on "enable "with X():" semantics for torch dispatch and function mode"


Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

[ghstack-poisoned]

samdow commented on torch/overrides.py:
```diff
@@ -1819,6 +1817,16 @@ def __init__(self):
         raise NotImplementedError()
 
+    def __enter__(self):
+        if hasattr(self, "inner"):
```
Flagging again that in the next PR in the stack, we remove the ability to set "inner" through the constructor which limits the errors that can happen here. In hindsight, I like that we set inner here instead of in the constructor because before this the following code is valid:

```
inner_mode = MyMode(inner=None)
with enable_torch_dispatch_mode(MyMode(inner=inner_mode)):
  ...
```

which is really weird because it will dispatch to `inner_mode` even though it's never used in a context manager. By setting it in enter, users will no longer be able to have this in their code

ezyang replied:
This is actually intentional; in functorch there are circumstances (custom vjp) where you close over some mode state and want to reapply it at some later point: you want to apply all the modes here. So the raw setter API let you apply as many modes as you want at once. I have never personally used it for anything though!

ezyang replied:
I think the code as written here is correct however! __enter__ should not be OK for restoring a mode; we can have a dedicated method for that

samdow commented on torch/overrides.py:
```diff
@@ -1819,6 +1817,16 @@ def __init__(self):
+    def __enter__(self):
+        if hasattr(self, "inner"):
+            raise RuntimeError(f"{self} has already been used as a mode, please create and use a fresh version")
+        old = _get_torch_function_mode()
```
@ezyang In the old `push_torch_function_mode`, if old was None, we would make the inner be `BaseTorchFunctionMode` whose torch function just called the original function with the original args. Do you remember why?

ezyang replied:
Yes it is so that self.inner.__torch_dispatch__() always worked. But since we strip off the original mode before calling into the function you can do func(*args, **kwargs) so supporting direct inner calls is much less pressing so I'm OK with dropping base mode (this is good for efficiency too, that base mode goops up stack frames).

Commit:
Update on "enable "with X():" behavior for torch dispatch and function mode"


Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

[ghstack-poisoned]

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -974,16 +974,75 @@ def test_push_mode_instance_errors(self):
 
-    def test_missing_inner_mode_ctor(self):
-        self.assertRaisesRegex(TypeError, 'push_torch_dispatch_mode', lambda: TorchDispatchMode())
+    def test_ctor_no_inner(self):
+        class A(TorchDispatchMode):
+            def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                return torch.zeros([])
+
+        with enable_torch_dispatch_mode(A()):
+            x = torch.randn((3, 4))
+
+        self.assertEqual(x, torch.zeros([]))
+
+    def test_with_mode(self):
+        class ErrorA(RuntimeError):
+            pass
+
+        class A(TorchDispatchMode):
+            def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                raise ErrorA()
+
+        with self.assertRaises(ErrorA):
+            with A():
+                torch.empty([])
+
+    def test_with_mode_created_separately(self):
+        class ErrorA(RuntimeError):
+            pass
+
+        class A(TorchDispatchMode):
+            def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                raise ErrorA()
+
+        x = A()
+        with self.assertRaises(ErrorA):
+            with x:
+                torch.empty([])
+
+    def test_with_nested_modes(self):
+        class ErrorA(RuntimeError):
```
nit: I kind of wish we could deduplicate these tests between torch_function mode and torch_dispatch mode. Not a big deal

zou3519 commented on torch/overrides.py:
```diff
@@ -1739,7 +1739,12 @@ def is_tensor_like(inp):
     @functools.wraps(f)
     def wrapped(self, *args, **kwargs):
-        with enable_torch_function_mode(self.inner):
+        if hasattr(self, "inner"):
+            inner = self.inner
+        else:
+            inner = None
```
nit: `getattr(self, "inner", default=None)`

zou3519 commented on torch/overrides.py:
```diff
@@ -1739,7 +1739,12 @@ def is_tensor_like(inp):
+        else:
+            inner = None
+
+        with enable_torch_function_mode(inner):
```
I remember we were debugging something related to inner being BaseMode or something last week, and that this PR should fix it. If so, then that would be worth adding as a test case

zou3519 commented on torch/utils/_mode_utils.py:
```diff
@@ -10,28 +10,15 @@
-                f"directly, pass the constructor to push_{meta_init_error_info.mode_name}_mode"
-            )
-        self.inner = inner
+        if inner is not undef:
+            self.inner = inner
```
No action required, but I have questions:

I missed this, but why is `undef = object()`?

Also, why not set self.inner = None if inner is undef?

samdow replied:
For both, we wanted to differentiate between:
`self.inner = None` which means "I am the outermost mode. Everything after me should redispatch without the mode stack"
`self.inner` isn't set yet which previously meant an error because it meant the user was trying to create a mode not using push or enable. Now it means "I'm a mode and my inner will be set in the future (specifically when entered through a context manager)." Concretely,

```
x = SomeMode()  # no inner
y = SomeMode() # no inner
with y: # y's inner is now None
  with x: # works! x's inner is now y
    ...

z = SomeMode(inner = None)
with y: # works because y's inner is None and it was originally None too
  with z: # errors because z's inner is None but the current mode is not None
    ...
```

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -65,7 +65,12 @@ def enable_torch_dispatch_mode(mode, *, replace=None, ignore_preexisting=False)
     @functools.wraps(f)
     def wrapped(self, *args, **kwargs):
-        with enable_torch_dispatch_mode(self.inner):
+        if hasattr(self, "inner"):
+            inner = self.inner
+        else:
+            inner = None
```
nit: getattr(self, "inner", default=None). This logic looks a lot like the one in torch_function :P

zou3519 reviewed (approved):
LGTM as well, minor nits and questions

Commit:
Update on "enable "with X():" behavior for torch dispatch and function mode"


Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

[ghstack-poisoned]

Commit:
Update on "enable "with X():" behavior for torch dispatch and function mode"


Allows users to write `with MyMode():`

Because the mode can be constructed separately from being used in a with statement, this required some hackery of the inner element. Specifically, inner is now being set on `__enter__` for with instead of in the constructor like with `push_X_mode`. This might cause some problems, but at least there's a nice error message. In the next PR in the stack, I remove the `push_X_mode` version which also means that we can remove the inner kwarg from the constructor, meaning that we only instantiate inner in enter

[ghstack-poisoned]

samdow commented:
@pytorchbot merge -g

samdow commented:
@pytorchbot merge -g
